File: AutoUpdate/AutoUpdate/Services/UserDataManager.cs
using AutoUpdate.Configuration;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;

namespace AutoUpdate.Services;

/// <summary>
/// Implementation of IUserDataManager for backing up and restoring user data.
/// </summary>
public class UserDataManager : IUserDataManager
{
    private readonly AutoUpdateOptions _options;
    private readonly ILogger<UserDataManager> _logger;

    public UserDataManager(
        IOptions<AutoUpdateOptions> options,
        ILogger<UserDataManager> logger)
    {
        _options = options.Value;
        _logger = logger;
    }

    /// <inheritdoc/>
    public async Task BackupAsync(string[] dataPaths, string backupPath, CancellationToken cancellationToken = default)
    {
        try
        {
            _logger.LogInformation("Starting user data backup to: {BackupPath}", backupPath);

            // Create backup directory with timestamp
            var timestamp = DateTimeOffset.Now.ToUnixTimeSeconds();
            var fullBackupPath = Path.Combine(backupPath, $"backup_{timestamp}");
            Directory.CreateDirectory(fullBackupPath);

            foreach (var dataPath in dataPaths)
            {
                cancellationToken.ThrowIfCancellationRequested();

                if (!Directory.Exists(dataPath) && !File.Exists(dataPath))
                {
                    _logger.LogWarning("Data path does not exist: {Path}", dataPath);
                    continue;
                }

                var dataName = Path.GetFileName(dataPath);
                var backupItemPath = Path.Combine(fullBackupPath, dataName);

                if (Directory.Exists(dataPath))
                {
                    // Copy directory
                    await CopyDirectoryAsync(dataPath, backupItemPath, cancellationToken);
                    _logger.LogInformation("Backed up directory: {Source} -> {Target}", dataPath, backupItemPath);
                }
                else if (File.Exists(dataPath))
                {
                    // Copy file
                    var targetDir = Path.GetDirectoryName(backupItemPath);
                    if (!string.IsNullOrEmpty(targetDir))
                    {
                        Directory.CreateDirectory(targetDir);
                    }

                    File.Copy(dataPath, backupItemPath, true);
                    _logger.LogInformation("Backed up file: {Source} -> {Target}", dataPath, backupItemPath);
                }
            }

            _logger.LogInformation("User data backup completed successfully");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error during user data backup");
            throw;
        }
    }

    /// <inheritdoc/>
    public async Task RestoreAsync(string backupPath, string[] dataPaths, CancellationToken cancellationToken = default)
    {
        try
        {
            _logger.LogInformation("Starting user data restore from: {BackupPath}", backupPath);

            // Find the most recent backup
            var backupDirs = Directory.GetDirectories(backupPath, "backup_*")
                .OrderByDescending(d => d)
                .ToArray();

            if (backupDirs.Length == 0)
            {
                _logger.LogWarning("No backup directories found");
                return;
            }

            var latestBackup = backupDirs[0];
            _logger.LogInformation("Using latest backup: {Backup}", latestBackup);

            foreach (var dataPath in dataPaths)
            {
                cancellationToken.ThrowIfCancellationRequested();

                var dataName = Path.GetFileName(dataPath);
                var backupItemPath = Path.Combine(latestBackup, dataName);

                if (!Directory.Exists(backupItemPath) && !File.Exists(backupItemPath))
                {
                    _logger.LogWarning("Backup item does not exist: {Path}", backupItemPath);
                    continue;
                }

                if (Directory.Exists(backupItemPath))
                {
                    // Remove existing directory if it exists
                    if (Directory.Exists(dataPath))
                    {
                        Directory.Delete(dataPath, true);
                    }

                    // Restore directory
                    await CopyDirectoryAsync(backupItemPath, dataPath, cancellationToken);
                    _logger.LogInformation("Restored directory: {Source} -> {Target}", backupItemPath, dataPath);
                }
                else if (File.Exists(backupItemPath))
                {
                    // Ensure target directory exists
                    var targetDir = Path.GetDirectoryName(dataPath);
                    if (!string.IsNullOrEmpty(targetDir))
                    {
                        Directory.CreateDirectory(targetDir);
                    }

                    // Restore file
                    File.Copy(backupItemPath, dataPath, true);
                    _logger.LogInformation("Restored file: {Source} -> {Target}", backupItemPath, dataPath);
                }
            }

            // Clean up backup directory
            try
            {
                Directory.Delete(latestBackup, true);
                _logger.LogInformation("Backup directory cleaned up: {Path}", latestBackup);
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "Failed to clean up backup directory: {Path}", latestBackup);
            }

            _logger.LogInformation("User data restore completed successfully");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error during user data restore");
            throw;
        }
    }

    private async Task CopyDirectoryAsync(string sourceDir, string targetDir, CancellationToken cancellationToken)
    {
        // Create target directory
        Directory.CreateDirectory(targetDir);

        // Copy all files and subdirectories
        foreach (var file in Directory.GetFiles(sourceDir, "*", SearchOption.AllDirectories))
        {
            cancellationToken.ThrowIfCancellationRequested();

            var relativePath = Path.GetRelativePath(sourceDir, file);
            var targetPath = Path.Combine(targetDir, relativePath);

            var targetFileDir = Path.GetDirectoryName(targetPath);
            if (!string.IsNullOrEmpty(targetFileDir))
            {
                Directory.CreateDirectory(targetFileDir);
            }

            File.Copy(file, targetPath, true);
        }
    }
}

File: AutoUpdateJS/src/services/UserDataManager.js
import path from 'path';
import { copyDirectory, ensureDirectory, remove, directoryExists } from '../utils/fileUtils.js';
import logger from '../utils/logger.js';

/**
 * Service for managing user data backup and restore operations.
 */
export class UserDataManager {
  /**
   * Creates a new UserDataManager instance.
   * @param {AutoUpdateConfig} config - Configuration instance
   */
  constructor(config) {
    this.config = config;
  }

  /**
   * Backs up user data to a timestamped directory.
   * @param {string[]} dataPaths - Paths to user data directories/files
   * @param {string} backupPath - Base backup directory path
   * @returns {Promise<string>} Path to the created backup directory
   */
  async backup(dataPaths, backupPath) {
    const timestamp = Date.now();
    const fullBackupPath = path.join(backupPath, `backup_${timestamp}`);

    try {
      logger.info(`Starting user data backup to: ${fullBackupPath}`);

      await ensureDirectory(fullBackupPath);

      for (const dataPath of dataPaths) {
        if (!(await directoryExists(dataPath))) {
          logger.warn(`Data path does not exist: ${dataPath}`);
          continue;
        }

        const dataName = path.basename(dataPath);
        const backupItemPath = path.join(fullBackupPath, dataName);

        try {
          await copyDirectory(dataPath, backupItemPath);
          logger.info(`Backed up directory: ${dataPath} -> ${backupItemPath}`);
        } catch (error) {
          logger.warn(`Failed to backup ${dataPath}: ${error.message}`);
        }
      }

      logger.info('User data backup completed successfully');
      return fullBackupPath;
    } catch (error) {
      logger.error(`Error during user data backup: ${error.message}`, error);
      throw error;
    }
  }

  /**
   * Restores user data from the most recent backup.
   * @param {string} backupPath - Base backup directory path
   * @param {string[]} dataPaths - Paths where to restore the data
   * @returns {Promise<void>}
   */
  async restore(backupPath, dataPaths) {
    try {
      logger.info(`Starting user data restore from: ${backupPath}`);

      // Find the most recent backup
      const fs = await import('fs/promises');
      let backupDirs = [];

      try {
        const entries = await fs.readdir(backupPath);
        backupDirs = entries
          .filter(entry => entry.startsWith('backup_'))
          .map(entry => path.join(backupPath, entry))
          .sort()
          .reverse(); // Most recent first
      } catch (error) {
        logger.warn(`No backup directory found at ${backupPath}`);
        return;
      }

      if (backupDirs.length === 0) {
        logger.warn('No backup directories found');
        return;
      }

      const latestBackup = backupDirs[0];
      logger.info(`Using latest backup: ${latestBackup}`);

      for (const dataPath of dataPaths) {
        const dataName = path.basename(dataPath);
        const backupItemPath = path.join(latestBackup, dataName);

        if (!(await directoryExists(backupItemPath))) {
          logger.warn(`Backup item does not exist: ${backupItemPath}`);
          continue;
        }

        try {
          // Remove existing directory if it exists
          if (await directoryExists(dataPath)) {
            await remove(dataPath);
          }

          // Restore from backup
          await copyDirectory(backupItemPath, dataPath);
          logger.info(`Restored directory: ${backupItemPath} -> ${dataPath}`);
        } catch (error) {
          logger.warn(`Failed to restore ${dataPath}: ${error.message}`);
        }
      }

      // Clean up backup directory
      try {
        await remove(latestBackup);
        logger.info(`Backup directory cleaned up: ${latestBackup}`);
      } catch (error) {
        logger.warn(`Failed to clean up backup directory: ${error.message}`);
      }

      logger.info('User data restore completed successfully');
    } catch (error) {
      logger.error(`Error during user data restore: ${error.message}`, error);
      throw error;
    }
  }

  /**
   * Backs up user data using the configured paths.
   * @returns {Promise<string>} Path to the created backup directory
   */
  async backupUserData() {
    return this.backup(this.config.userDataPaths, this.config.backupDirectory);
  }

  /**
   * Restores user data using the configured paths.
   * @returns {Promise<void>}
   */
  async restoreUserData() {
    return this.restore(this.config.backupDirectory, this.config.userDataPaths);
  }

  /**
   * Cleans up old backup directories, keeping only the most recent ones.
   * @param {number} keepCount - Number of recent backups to keep (default: 5)
   * @returns {Promise<void>}
   */
  async cleanupOldBackups(keepCount = 5) {
    try {
      const fs = await import('fs/promises');

      if (!(await directoryExists(this.config.backupDirectory))) {
        return;
      }

      const entries = await fs.readdir(this.config.backupDirectory);
      const backupDirs = entries
        .filter(entry => entry.startsWith('backup_'))
        .map(entry => ({
          name: entry,
          path: path.join(this.config.backupDirectory, entry),
          timestamp: parseInt(entry.replace('backup_', ''))
        }))
        .sort((a, b) => b.timestamp - a.timestamp); // Most recent first

      if (backupDirs.length <= keepCount) {
        return;
      }

      const dirsToRemove = backupDirs.slice(keepCount);

      for (const dir of dirsToRemove) {
        try {
          await remove(dir.path);
          logger.info(`Cleaned up old backup: ${dir.name}`);
        } catch (error) {
          logger.warn(`Failed to clean up backup ${dir.name}: ${error.message}`);
        }
      }

      logger.info(`Cleaned up ${dirsToRemove.length} old backup(s)`);
    } catch (error) {
      logger.error(`Error during backup cleanup: ${error.message}`, error);
    }
  }
}
